File: app/src/main/java/com/madebyaron/betterhealth/viewmodel/StepsModel.java
package com.madebyaron.betterhealth.viewmodel;

import android.app.Application;
import android.content.Context;
import android.content.Intent;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;
import com.madebyaron.betterhealth.stepcounter.StepCounterService;
import com.madebyaron.betterhealth.database.StepsDB;
import java.util.*;
import java.util.concurrent.TimeUnit;

public class StepsModel extends AndroidViewModel {

	private StepsDB stepsApi = new StepsDB(getApplication());

	private long refreshInterval = 1000L;
	private Timer refreshTimer = null;

	public MutableLiveData<String> steps = new MutableLiveData<>();
	public MutableLiveData<Boolean> showStartButton = new MutableLiveData<>();
	public MutableLiveData<Boolean> showStopButton = new MutableLiveData<>();

	public StepsModel(Application application) {
		super(application);
		onServiceRunning(StepCounterService.isRunning());
		StepCounterService.addRunningListener(this::onServiceRunning);
	}

	@Override
	protected void onCleared() {
		StepCounterService.removeRunningListener(this::onServiceRunning);
	}

	public void startStepCounter() {
		Context context = getApplication();
		context.startService(new Intent(context, StepCounterService.class));
	}

	public void stopStepCounter() {
		Context context = getApplication();
		context.stopService(new Intent(context, StepCounterService.class));
	}

	public void startRefreshing() {
		stopRefreshing();

		refreshTimer = new Timer();
		refreshTimer.scheduleAtFixedRate(new TimerTask() {

			@Override
			public void run() {
				Date from = new Date();
				from.setTime(from.getTime() - TimeUnit.DAYS.toMillis(1));
				stepsApi.getStepCount(
						from, new Date(),
						stepCount -> steps.postValue(stepCount.toString())
				);
			}
		}, 0, refreshInterval);
	}

	public void stopRefreshing() {
		if (refreshTimer != null) refreshTimer.cancel();
		refreshTimer = null;
	}

	private void onServiceRunning(boolean running) {
		showStartButton.postValue(!running);
		showStopButton.postValue(running);
	}
}

File: app/src/main/java/com/madebyaron/betterhealth/viewmodel/StepsModel.kt
package com.madebyaron.betterhealth.viewmodel

import android.app.Application
import android.content.Context
import android.content.Intent
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import com.madebyaron.betterhealth.stepcounter.StepCounterService
import com.madebyaron.betterhealth.database.StepsDB
import java.util.*
import java.util.concurrent.TimeUnit

class StepsModel(application: Application) : AndroidViewModel(application) {

	private val context: Context
		get() = getApplication()

	private val stepsApi = StepsDB(context)

	private val refreshInterval = 1000L
	private var refreshTimer: Timer? = null

	val steps = MutableLiveData<String>()
	val showStartButton = MutableLiveData<Boolean>()
	val showStopButton = MutableLiveData<Boolean>()

	init {
		onServiceRunning(StepCounterService.running)
		StepCounterService.addRunningListener(this::onServiceRunning)
	}

	override fun onCleared() {
		StepCounterService.removeRunningListener(this::onServiceRunning)
	}

	fun startStepCounter() {
		context.startService(Intent(context, StepCounterService::class.java))
	}

	fun stopStepCounter() {
		context.stopService(Intent(context, StepCounterService::class.java))
	}

	fun startRefreshing() {
		stopRefreshing()

		refreshTimer = Timer()
		refreshTimer?.scheduleAtFixedRate(object : TimerTask() {
			override fun run() {
				val from = Date()
				from.time -= TimeUnit.DAYS.toMillis(1)
				stepsApi.getStepCount(from, Date()) { steps.postValue(it.toString()) }
			}
		}, 0, refreshInterval)
	}

	fun stopRefreshing() {
		refreshTimer?.cancel()
		refreshTimer = null
	}

	private fun onServiceRunning(running: Boolean) {
		showStartButton.postValue(!running)
		showStopButton.postValue(running)
	}
}